### ISSUE TITLE
Crash in ViewDependentState

### ISSUE THREAD

**Issue by OpenSiege** [vsg-dev/VulkanSceneGraph#400]
**Describe the bug**
After update my project to use master, I am getting a crash in my local application. Looks like the culprit is here: https://github.com/vsg-dev/VulkanSceneGraph/blob/master/src/vsg/state/ViewDependentState.cpp#L124

My debugger is reporting that ```bufferedDescriptor.size()``` is 0.

**To Reproduce**
Running vsg-siege with command line option ```--state "RegionTestState```.

**Expected behavior**
Not to crash

**Desktop (please complete the following information):**
 - OS: Windows
 - Version: 2019

**Additional context**
Stack trace:

```
>	vsg-siege.exe!vsg::ViewDependentState::clear() Line 124	C++
 	vsg-siege.exe!vsg::RecordTraversal::apply(const vsg::View & view) Line 434	C++
 	vsg-siege.exe!vsg::View::t_accept<vsg::View const ,vsg::RecordTraversal>(const vsg::View & node, vsg::RecordTraversal & visitor) Line 42	C++
 	vsg-siege.exe!vsg::View::accept(vsg::RecordTraversal & visitor) Line 47	C++
 	vsg-siege.exe!vsg::Group::t_traverse<vsg::Group const ,vsg::RecordTraversal>(const vsg::Group & node, vsg::RecordTraversal & visitor) Line 39	C++
 	vsg-siege.exe!vsg::Group::traverse(vsg::RecordTraversal & visitor) Line 44	C++
 	vsg-siege.exe!vsg::RenderGraph::accept(vsg::RecordTraversal & recordTraversal) Line 137	C++
 	vsg-siege.exe!vsg::Group::t_traverse<vsg::Group const ,vsg::RecordTraversal>(const vsg::Group & node, vsg::RecordTraversal & visitor) Line 39	C++
 	vsg-siege.exe!vsg::Group::traverse(vsg::RecordTraversal & visitor) Line 44	C++
 	vsg-siege.exe!vsg::RecordTraversal::apply(const vsg::Group & group) Line 126	C++
 	vsg-siege.exe!vsg::Inherit<vsg::Group,vsg::CommandGraph>::accept(vsg::RecordTraversal & visitor) Line 83	C++
 	vsg-siege.exe!vsg::CommandGraph::record(std::vector<vsg::ref_ptr<vsg::CommandBuffer>,std::allocator<vsg::ref_ptr<vsg::CommandBuffer>>> & recordedCommandBuffers, vsg::ref_ptr<vsg::FrameStamp> frameStamp, vsg::ref_ptr<vsg::DatabasePager> databasePager) Line 153	C++
 	vsg-siege.exe!vsg::RecordAndSubmitTask::record(std::vector<vsg::ref_ptr<vsg::CommandBuffer>,std::allocator<vsg::ref_ptr<vsg::CommandBuffer>>> & recordedCommandBuffers, vsg::ref_ptr<vsg::FrameStamp> frameStamp) Line 81	C++
 	vsg-siege.exe!vsg::RecordAndSubmitTask::submit(vsg::ref_ptr<vsg::FrameStamp> frameStamp) Line 59	C++
 	vsg-siege.exe!vsg::Viewer::recordAndSubmit() Line 621	C++
 	vsg-siege.exe!ehb::Game::exec() Line 93	C++
 	vsg-siege.exe!main(int argc, char * * argv) Line 38	C++
 	[External Code]	

```

**Comment by robertosfield** [comment]
It looks like VisualStudio has a problem with % 0.  I've rewritten the bufferIndex to handle this case.  

The fix is checked in : df8b21515e95b9f2171235434b8577adc7dc9935

Could you test it on your system and let me know if it's resolved?

**Comment by OpenSiege** [comment]
Looks like it's moved down into the

```
void ViewDependentState::copy()
{
    //    std::cout<<"ViewDependentState::copy()"<<std::endl;

    auto& descriptorData = bufferedDescriptors[bufferIndex];
    for (auto& bufferInfo : descriptorData.lightDescriptor->bufferInfoList)
    {
        bufferInfo->copyDataToBuffer();
    }
}
```

I get an out of index error on the below since bufferedDescriptors is empty

```
auto& descriptorData = bufferedDescriptors[bufferIndex];
```

**Comment by robertosfield** [comment]
Does this happen with vsgviewer or vsgligts?

It's curious as it the ViewDependentState::compile(..) should have already been called an initialized the bufferedDescriptors container.

What is the size of bufferDesctiptors and bufferIndex?

**Comment by OpenSiege** [comment]
> Does this happen with vsgviewer or vsgligts?

No, it doesn't.

> What is the size of bufferDesctiptors and bufferIndex?

```
bufferedDescriptors = { size=0 }
bufferIndex = 0
```
I have this sneaking suspicion it's related to [vsgExamples/#124](https://github.com/vsg-dev/vsgExamples/issues/124). When I take the test program linked in that issue [builder.cpp.txt](https://github.com/vsg-dev/VulkanSceneGraph/files/7990507/builder.cpp.txt) it also crashes with the following:

```
 	nvoglv64.dll!00007ff8075adfde()	Unknown
>	vsg-siege.exe!vsg::PipelineLayout::Implementation::Implementation(vsg::Device * device, const std::vector<vsg::ref_ptr<vsg::DescriptorSetLayout>,std::allocator<vsg::ref_ptr<vsg::DescriptorSetLayout>>> & descriptorSetLayouts, const std::vector<VkPushConstantRange,std::allocator<VkPushConstantRange>> & pushConstantRanges, unsigned int flags) Line 116	C++
 	vsg-siege.exe!vsg::Inherit<vsg::Object,vsg::PipelineLayout::Implementation>::create<vsg::ref_ptr<vsg::Device> &,std::vector<vsg::ref_ptr<vsg::DescriptorSetLayout>,std::allocator<vsg::ref_ptr<vsg::DescriptorSetLayout>>> &,std::vector<VkPushConstantRange,std::allocator<VkPushConstantRange>> &,unsigned int &>(vsg::ref_ptr<vsg::Device> & <args_0>, std::vector<vsg::ref_ptr<vsg::DescriptorSetLayout>,std::allocator<vsg::ref_ptr<vsg::DescriptorSetLayout>>> & <args_1>, std::vector<VkPushConstantRange,std::allocator<VkPushConstantRange>> & <args_2>, unsigned int & <args_3>) Line 66	C++
 	vsg-siege.exe!vsg::PipelineLayout::compile(vsg::Context & context) Line 90	C++
 	vsg-siege.exe!vsg::GraphicsPipeline::compile(vsg::Context & context) Line 136	C++
 	vsg-siege.exe!vsg::BindGraphicsPipeline::compile(vsg::Context & context) Line 247	C++
 	vsg-siege.exe!vsg::StateGroup::compile(vsg::Context & context) Line 81	C++
 	vsg-siege.exe!vsg::CompileTraversal::apply(vsg::StateGroup & stateGroup) Line 86	C++
 	vsg-siege.exe!vsg::Inherit<vsg::Group,vsg::StateGroup>::accept(vsg::Visitor & visitor) Line 81	C++
 	vsg-siege.exe!vsg::Builder::compile(vsg::ref_ptr<vsg::Node> subgraph) Line 250	C++
 	vsg-siege.exe!vsg::Builder::createBox(const vsg::GeometryInfo & info, const vsg::StateInfo & stateInfo) Line 434	C++
 	vsg-siege.exe!builderTestAfterCompile(vsg::ref_ptr<vsg::Builder> builder, vsg::ref_ptr<vsg::Group> scene) Line 103	C++
 	vsg-siege.exe!main(int argc, char * * argv) Line 182	C++
 	[External Code]	
```

Not 100% sure if it's related but it does crash.

**Comment by robertosfield** [comment]
Hi Sam,

Thanks for the example and explanation.  This has given me enough clues to
look for existing vsgexamples that fail with the usage case you highlight -
the vsgintersection example is now crashing.

I don't know yet the cause but now I've got an example that highlights a
problem usage case I can dig deeper.

Cheers.
Robert.

**Comment by OpenSiege** [comment]
Maybe it’s worthwhile expanding out a headless example as a quick unit test for merging things into the graph during runtime? It might help pinpointing these types of crashes along the development path.

**Comment by robertosfield** [comment]
On Thu, 3 Feb 2022 at 08:31, OpenSiege ***@***.***> wrote:

> Maybe it’s worthwhile expanding out a headless example as a quick unit
> test for merging things into the graph during runtime? It might help
> pinpointing these types of crashes also
>
I don't think being headless is a particular help.  Having examples that
add scene graph elements dynamically are useful, there isn't just one usage
case for adding data so it'd need to be several examples.

**Comment by robertosfield** [comment]
Today I'm working on a client work, once that's complete I'll jump back to
investigating this bug.

**Comment by OpenSiege** [comment]
Sorry, should have been more clear. Making them headless would allow GitHub actions to run them during workflows.

**Comment by robertosfield** [comment]
On Thu, 3 Feb 2022 at 08:44, OpenSiege ***@***.***> wrote:

> Sorry, should have been more clear. Making them headless would allow
> GitHub actions to run them during workflows.
>
> I understand that, but you can only catch some types of errors with
headless tests.

**Comment by robertosfield** [comment]
I have got vsgintersection working and checked in a fix to VSG master.  The
issue with vsgintersection is that the local compile traversal used by
vsg::Builder doesn't cache a ViewDependentState as it is not locally set.
I'll need to come back to this.  Can you try VSG master on you test program?

**Comment by OpenSiege** [comment]
Same crash and stack trace on both my siege code and the builder example.

**Comment by robertosfield** [comment]
On Thu, 3 Feb 2022 at 19:09, OpenSiege ***@***.***> wrote:

> Same crash and stack trace on both my siege code and the builder example.
>
Oh crqp. Sorry to hear it's still broken.  I'm still working on client work
and should finish it tomorrow then can do a deeper dive.

Short term revert to before the merge of the Lights branch.

**Comment by robertosfield** [comment]
I've finally wrapped up the RenderPass2 work that took me away from resolve
Builder compile bug.  It's near end of working day here so I'll need to
return to investigate the issue tomorrow.

What I can say I have figured out the how to reproduce the problem usage
case with the vsgintersection example.  It doesn't crash if you run it with
no model on the command line, but if you run vsgintersection with a model
like models/lz.vsgt then it crashes as soon as you try to insert a model.

My interpretation of this problem usage case is that invoking vsg::Builder
during the main loo[/after the initial viewer.compile() is that the
vsg::Builder compile doesn't know about the View and it's
ViewDependentState so doesn't pull the associated ViewDiscriptorSetLayout
and ViewBindDescriptorSet's in and doesn't make sure they are compiled.

Tomorrow my first line of investigation will be look at how to pass the
View state into the vsg::Builder compile.

**Comment by robertosfield** [comment]
I have implemented a temporary fix that passes the View to the vsg::Context/Builder constructors so the compile traversal can utilize the appropriate ViewDependentState.  The is checked into a new [ContextExperiment](https://github.com/vsg-dev/VulkanSceneGraph/tree/ContextExperiment) branch.

I have update the vsgintersection example to pass the View to the Build for compiles, with this in place it's now possible load a 3d model then utilize the vsg::Builder at runtime without the crash.  This also should make it possible to use the CompileTraversal more generally.  The required changes to vsgintersection are checked into vsgExamples [ContextExperiment](https://github.com/vsg-dev/vsgExamples/tree/ContextExperiment) branch.

**Comment by robertosfield** [comment]
I view the ContextExperiment branches of VSG and vsgExamples are proof of concept, not a final solution.

The final solution will require more extensive changes to CompileTraversal/Context/Builder and classes that use them so that they can handle a wide range of usage cases.  Exactly what form this will take I am working on.

**Comment by robertosfield** [comment]
I have now completed much of the [CompileRefactor](https://github.com/vsg-dev/VulkanSceneGraph/tree/CompileRefactor) work and have updated the vsgExamples vsgintersection example to utilize the new functionality.  This fixes the crash and make the ContextExperiment redundant so I've deleted this branch. 

The required vsgintersection code changes can be found in the vsgExamples [CompileRefactor](https://github.com/vsg-dev/VulkanSceneGraph/tree/CompileRefactor) branch.  My plan is to provide easier support for setting up the vsg::CompileTraversal that needs to be assigned to vsg::Builder for updates at runtime, but for now it is at least funcitonal, even if it's a bit more long winded to set up then it will be once the CompileRefactor work is completed at the end of next week.

**Comment by sbrkopac** [comment]
So to get my app to crashing, I'll have to follow the same methodology described in your vsgExamples fix?

**Comment by robertosfield** [comment]
Yes just follow what vsgintersection is doing in the CompilrTefactor branch.

**Comment by sbrkopac** [comment]
Looks like this fixes my example program. Just to clarify - in your example, you aren't using ```createCommmandGraphForView`` because you have to manually pass the view into the ```CompileTraversal```?

**Comment by robertosfield** [comment]
Yes, direct creation of the vsg::View rather calling createCommandGraphForView() is needed so you know which VIew to pass onto the CompileTraversal. 

I will come up with a convenience function to either configures or creates a CompileTraversal to compiles for all the CommandGraph's assigned to a VIewer.  Once this is available you'll be able to just use the createCommandGraphForView as before and rely upon this extra function to create things for you.

Another alternative is to cache such a CompileTraversal in the Viewer itself

**Comment by sbrkopac** [comment]
Just for sake of reference, currently a user has to call compile() on their viewer shortly after setup if they are doing any dynamic scenegraph type work - otherwise the descriptor buffers don't get setup properly and it results in a crash.

**Comment by robertosfield** [comment]
I have just checked in additional checks to ViewDependentState::copy() and ViewDependentState::bindDescriptorSets() to handle the case when the ViewDependentState::compile() hasn't yet been called.  Could you test it out?  The fix is merged with master e03cc45fe4619d91e81f4394504c4f3d5311a5aa

**Comment by sbrkopac** [comment]
Yup, this fixes the remaining crash.
